modules/tensor_mechanics/src/materials/ADPowerLawCreepStressUpdate.C

```diff
@@ -58,12 +58,13 @@ ADPowerLawCreepStressUpdate::computeStressInitialize(const ADReal & /*effective_
   _exp_time = std::pow(_t - _start_time, _m_exponent);
 }
 
-ADReal
-ADPowerLawCreepStressUpdate::computeResidual(const ADReal & effective_trial_stress,
-                                             const ADReal & scalar)
+template <typename ScalarType>
+ScalarType
+ADPowerLawCreepStressUpdate::computeResidualInternal(const ADReal & effective_trial_stress,
+                                                     const ScalarType & scalar)
 {
-  const ADReal stress_delta = effective_trial_stress - _three_shear_modulus * scalar;
-  const ADReal creep_rate =
+  const ScalarType stress_delta = effective_trial_stress - _three_shear_modulus * scalar;
+  const ScalarType creep_rate =
       _coefficient * std::pow(stress_delta, _n_exponent) * _exponential * _exp_time;
   return creep_rate * _dt - scalar;
 }
```

Comment:
This is what the end users need to do if they want to convert their stress update classes to use AD return mapping derivative.

Cheers!

Reply:
🍺

Reply:
AD types in this method:
- `_exponential`
- `effective_trial_stress`
- `_three_shear_modulus`

So without auto-type-promotion, this could would need to look something like
```c++
  auto make_scalar = [](const auto & down_type)
  {
    ScalarType ret;
    ret.value() = down_type;
    return ret;
  };

  const ScalarType stress_delta =
      make_scalar(effective_trial_stress) - make_scalar(_three_shear_modulus) * scalar;
  const ScalarType creep_rate =
      _coefficient * std::pow(stress_delta, _n_exponent) * make_scalar(_exponential) * _exp_time;
  return creep_rate * _dt - scalar;
```
Or would you do something differently @roystgnr ?

Reply:
We could probably create a global template function for this for which we only instantiate for `ChainedReal` and `ChainedADReal`. Then remove the `CompareTypes` from our framework headers